sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+  new (Storage) Foo;
+});
+----
+
+==== Interaction with `reqd_sub_group_size` attribute
+
+The `reqd_sub_group_size` attribute is a bit of a special case comparing to
+other optional kernel features, because it requires to compile a kernel in a
+certain way, which may require special handling for all functions which are
+called from it.
+
+When the same helper function is called from two or more kernels
+with different `reqd_sub_group_size` attribute, it may be required for the
+implementation to duplicate that helper function to create different versions
+of it tailored to different sub-group sizes. It can be done in a straightforward
+manner when operating on a static call graph.
+
+Virtual member functions are essentially called indirectly and pointers to them
+are initialized just once when an object of a polymorhpic class is being
+created. Therefore, to support calling such virtual member function from two or
+more kernels with different `reqd_sub_group_size`, each kernel may need to
+receive a different pointer to a different version of a virtual member function.
+
+To avoid possibly posing such multi-versioning requirements on implementations,
+virtual member functions can only be called from kernels with _primary_
+sub-group-size as defined by
+link:../proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc[
+sycl_ext_oneapi_named_sub_group_sizes] extension, or otherwise behavior is
+undefined.
+
+NOTE: for implementations that don't support
+`sycl_ext_oneapi_named_sub_group_sizes` extension, virtual member functions can
+only be called from kernels which *don't* have `reqd_sub_group_size` attribute
+set on them explicitly, or otherwise behavior is undefined.
+
+=== Kernel bundles and device images
+
+When an object of a polymorphic class is constructed, it stores a pointer to
+virtual table, which points to its virtual member functions. Addresses of those
+functions are accessible and valid only within a kernel bundle containing a
+kernel which used to construct an object.
+
+Performing calls to virtual member functions of an object constructed in a
+kernel from a different kernel bundle is an undefined behavior.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct Base {
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}
+};
+
+class Constructor;
+class Use;
+
+int main() {
+  sycl::queue Q;
+
+  Base *Obj = sycl::malloc_device<Base>(1, Q);
+  int *Result = sycl::malloc_shared<int>(2, Q);
+
+  auto bundleA
+    = sycl::get_kernel_bundle<sycl::bundle_state::executable>(Q.get_context(),
+        {sycl::get_kernel_id<Constructor>()});
+  auto bundleB
+    = sycl::get_kernel_bundle<sycl::bundle_state::executable>(Q.get_context(),
+        {sycl::get_kernel_id<Use>()});
+
+
+  Q.submit([&](sycl::handler &CGH) {
+    CGH.use_kernel_bundle(bundleA);
+    CGH.single_task<Constructor>([=] {
+      // Only placement new can be used within device functions.
+      new (Obj) Derived;
```

Comment:
Should the definition of `Derived` class be added to this code snippet?

Reply:
Right, it should be `Base`, fixed in 683bc66e92e3acb4c652791cfd7a815e5afa5783